Add a simple object pool based on a freelist, apply to `TemplateInstance`

andralex opened the pull request:
The `Pool` template allows recycling of memory allocated for specific types. It implements a simple freelist that dispenses objects with `make`. Objects can be returned to the pool with `dispose`. Objects that have been `dispose`d should not be used any longer.

This improves memory usage in a large app in my measurements, albeit only a little (1%). The trick here is to insert calls to `dispose` wherever appropriate. I only chose a handful of obvious cases, but I'm sure there are more.

andralex committed:
Add an object pool

ibuclaw commented:
If my memory serves correctly, valgrind said that this line is one of the biggest wasters of memory.

https://github.com/dlang/dmd/blob/8075f0ad31c6cab9d699a571ce086e86ea38af7a/src/dmd/dtemplate.d#L893

andralex commented:
@ibuclaw interesting, I see that only a few lines down `lastConstraint` may be assigned `null`. In that case we could free it. Not sure how frequently that happens, but definitely worth trying.

ibuclaw commented:
> @ibuclaw interesting, I see that only a few lines down `lastConstraint` may be assigned `null`. In that case we could free it. Not sure how frequently that happens, but definitely worth trying.

Template constraints fail far more than they succeed in a typical compilation, wouldn't want to put a ratio on that though.

Replacing syntaxCopy would be tricky though, as there's an indeterminate number of new and memcpy'ing done.

rainers commented:
Would be nice to see how -lowmem handles your large application and whether it is (negatively) affected by pooling memory which avoids it from being collected. I don't expect a notable impact, though.

UplinkCoder commented:
@andralex What is the impact on the time dimension?

UplinkCoder commented on src/dmd/dtemplate.d:
```diff
@@ -8266,3 +8281,29 @@ void printTemplateStats()
+struct Pool(T)
+if (is(T == class))
+{
+    private static T root;
```
Needs a comment.

UplinkCoder commented on src/dmd/dtemplate.d:
```diff
@@ -8266,3 +8281,29 @@ void printTemplateStats()
     }
 }
+
+struct Pool(T)
```
Should go into `root` or some other utility file, not in `dtemplate.d`

andralex replied:
OK

andralex commented:
> Would be nice to see how -lowmem handles your large application and whether it is (negatively) affected by pooling memory which avoids it from being collected. I don't expect a notable impact, though.

It's actually difficult to try that because there are many subprojects and couldn't find a way to propagate a flag to all. Freelists do prevent garbage from being collected, but just like you I assume the impact won't be measurable. One solution that works pretty nicely is to limit the length of the freelist.

andralex commented:
> @andralex What is the impact on the time dimension?

No measurable change for now.

NilsLankila commented on src/dmd/dtemplate.d:
```diff
@@ -8266,3 +8281,29 @@ void printTemplateStats()
+        *(cast(T*) goner) = root;
+        root = goner;
+        // Obliterate the memory so attempting to reuse the object after dispose is likely to fail
+        memset(cast(ubyte*) goner + (void*).sizeof, 0xf0, T.classinfo.initializer.length - (void*).sizeof);
```
is't a problem here would be that **pointer to** the disposed instances will think their target is still alive ?

andralex replied:
That's actually good - we don't want the GC to collect objects from the freelist.

dnadlinger commented:
> It's actually difficult to try that because there are many subprojects and couldn't find a way to propagate a flag to all.

Not being facetious here, but have you tried putting it in dmd.conf/environment variables?

andralex committed:
Move Pool to root.rmem

andralex commented:
> > It's actually difficult to try that because there are many subprojects and couldn't find a way to propagate a flag to all.
> 
> Not being facetious here, but have you tried putting it in dmd.conf/environment variables?

Putting it in the environment a la `DFLAGS="-lowmem" dub build` doesn't work, sadly. I didn't try the dmd.conf approach - thanks for the idea.

andralex committed:
Use Pool for TypeIdentifier, too

andralex commented:
Now `TypeIdentifier` uses `Pool`, too. I found a few opportunities to insert calls to `dispose`, most likely there are others. 1% improvement in memory use.

andralex committed:
One more dispose, also scope is better than pool

UplinkCoder commented:
`fail_compilation/fail14554.d`
Seems to fail because of a null ptr deref.

andralex committed:
Add debugging capability, fix crash

andralex commented:
@UplinkCoder thanks. I think it had to do with one too many `dispose` calls - the one in dtemplat.de line 3116. I removed that and added a simple tracing facility.

andralex committed:
__FILE__ and __LINE__ cannot be part of expressions in default paraemters, so drop that

andralex committed:
Change order of log lines so it's easy to use with simple tools such as sort

andralex commented:
@ibuclaw yah, your memory seems to jibe with my measurements. I just used the tracing in `Pool` on a few distinct projects and for all it seems the top places that allocate (fresh) memory regardless of project are (showing line numbers on current master):

* https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L5895
* https://github.com/dlang/dmd/blob/master/src/dmd/mtype.d#L5344
* https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L2969

It looks like recouping some of the allocations created in these places would be greatly beneficial.

The last place is also the one that gets to reuse most memory. Conversely, the most advantageous `dispose` is right before the `continue` here:

* https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L2978

andralex committed:
Comment the debug out

andralex committed:
Better function name

looked-at-me commented:
This save 3 mb from ~2 GB for my project. I don't think this PR is worth the complexity it is adding. Especially since it is doing effectively what the GC should already be doing.

Your probably better off looking into RTInfo in druntime. **Every** aggregate generates an instance of RTInfo which calls __traits getBitmapPointer or w/e it's called. It generates an array of bits for the type, then effectively discards that array to generate a worse AST representation using ArrayExp filled with IntegerExp. Since this happens for every type is is quite expensive and backwards. 

https://github.com/dlang/druntime/blob/master/src/object.d#L2956
https://github.com/dlang/dmd/blob/master/src/dmd/semantic3.d#L1473
https://github.com/dlang/dmd/blob/master/src/dmd/traits.d#L406

You'll probably save more than 3 mb spending your time removing that redundancy than duplicating work that's already done by the GC.

andralex commented:
> This saves 3 mb from ~2 GB for my project. I don't think this PR is worth the complexity it is adding.

This is just a foot in the door. The point is to insert more calls to dispose() in later PRs (the ones I inserted are fairly simple and ignorant; I'm sure people more familiar with the compiler codebase will find better opportunities).

> Especially since it is doing effectively what the GC should already be doing.

The compiler is not even using the GC. It's using a monotonic allocation scheme with no ability to reuse memory at all. The freelists offer the ability to reuse objects in a user-controlled fashion.

> Your probably better off looking into RTInfo in druntime. **Every** aggregate generates an instance of RTInfo which calls __traits getBitmapPointer or w/e it's called. A single type in my project causes over 60,000 allocations because of the way this is implemented. It generates an array of bits for the type, then effectively discards that array to generate a worse AST representation using ArrayExp filled with IntegerExp. Since this happens for every type it is quite expensive and backwards.
> 
> https://github.com/dlang/druntime/blob/master/src/object.d#L2956
> https://github.com/dlang/dmd/blob/master/src/dmd/semantic3.d#L1473
> https://github.com/dlang/dmd/blob/master/src/dmd/traits.d#L406

This is good stuff, thanks. Regardless of whether I can do something about it, I encourage you to take a look into improving the situation. Thanks!

WalterBright commented on src/dmd/root/rmem.d:
```diff
@@ -410,3 +410,78 @@ pure nothrow unittest
+    */
+    static void dispose(string f = __FILE__, uint l = __LINE__, A...)(T goner)
+    {
+        trace!("dispose", f, l)();
```
If `debug` is on, this should stomp on the object with `0xDEADBEEF`.

WalterBright replied:
Perhaps it should anyway, as that will prevent false pointers from being recognized by the GC.

andralex replied:
I changed the logic so the freelist is used ONLY if the GC is not used. In debug mode the memory is now stomped with `0xdeadbeef`.

UplinkCoder commented:
I would agree with @looked-at-me.
This PR solves nothing which `-lowmem` does not solve more effectively. 
It introduces a further element of human error due to manually placed dispose calls.

I am not sure that the human-error door is a door we want to get our foot into.
It could be justified if huge performance gains / memory reductions were expected.

So is this something you are expecting in the future?
And if so on what basis?

andralex committed:
Make the Pool active only if the GC is not used.

UplinkCoder commented:
At least in debug mode use-after-dispose will be visible now, that's something.

looked-at-me commented:
> This is just a foot in the door. The point is to insert more calls to dispose() in later PRs (the ones I inserted are fairly simple and ignorant; I'm sure people more familiar with the compiler codebase will find better opportunities).

Best case it'll do what the GC does, but it is very unlikely it will be able to reach that case. This method won't be able to catch the more complicated cases that the GC will. 

> The compiler is not even using the GC. It's using a monotonic allocation scheme with no ability to reuse memory at all. The freelists offer the ability to reuse objects in a user-controlled fashion.

That's a choice that was made so that it runs faster. If you are having memory problems then you can enable the GC. 

It's not really a problem of memory being allocated then not used, the way the frontend is built and operated on the AST is the problem. As with RTInfo it is being expanded into an AST instead of condensing to something closer to what the actual representation should be. Trying to optimize that will just bloat the AST with special edge cases that need to then be handled in every Vistior.

andralex commented:
> So are huge performance gains something you are expecting in the future?
> And if so on what basis?

Thanks for asking. Per the numbers published in the forum, dmd allocates preferentially a small number of types (below are the top ones up to a 10x drop in bytes allocated in a large project):

```
5308225728 27647009 192 dmd.mtype.TypeIdentifier
5295393792 13790088 384 dmd.dtemplate.TemplateInstance
2656326464 12709696 209 dmd.declaration.AliasDeclaration
2322240384 12095002 192 dmd.dsymbol.ScopeDsymbol
2254404450 7391490 305 dmd.declaration.VarDeclaration
1356347520 2055072 660 dmd.func.FuncDeclaration
1296538520 7046405 184 dmd.mtype.TypeInstance
691832155 3374791 205 dmd.mtype.TypeFunction
597424672 2872234 208 dmd.declaration.TupleDeclaration
540168864 11253518 48 dmd.mtype.Parameter
```

When allocations are concentrated in a few size classes, the go-to method is a freelist for those most frequent size classes fronting whatever allocator is used (see e.g. [this paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.96.4727&rep=rep1&type=pdf) although that is fronting a heap allocator for generality purposes).

The challenge, of course, is to find opportunities for `dispose()`ing objects for later reuse. The 1% gains (for that same project) for virtually no effort suggests there's good potential.

andralex commented:
For more information here are the cumulative percentages of the most allocated classes. The top 4 classes allocated contribute more than half of the memory allocated for all classes.

<img width="645" alt="Screen Shot 2020-07-23 at 9 05 42 PM" src="https://user-images.githubusercontent.com/566679/88352651-8dd2a680-cd28-11ea-90da-13c0fe7777cd.png">

looked-at-me commented:
What kind of project has 511,487 modules?

NilsLankila commented:
Awesome @looked-at-me, you seen this . Maybe it's a consequence of syntaxCopy() ?

UplinkCoder commented:
@looked-at-me one which has 13790088 Template instances. 
It comes to 26 template instances per module.
It's rather large :)

andralex commented:
Heh, interesting. Clearly there are more instances of `Module` than actual modules. Maybe seeing where that comes from may be of interest, too. The scale is impressive though - even if we got rid of all `Module`s, we'd save 1%.

andralex commented:
@NilsLankila here's a thought: if indeed `syntaxCopy` causes a lot of `Module`s to be unnecessarily copied (as I speculate they never change after creation), perhaps making `Module` and other such objects `immutable` members. Then it is obvious they don't need to be duplicated, and we have the compiler verify that for us.

NilsLankila commented:
That's just a guess. Possibly: referefenced Module would be copied. This would happen for example because an instance is not necessarily instanciated in the module where it is declared, but rather "pulled" in that module.

NilsLankila commented:
Or more simply that can be a Module member somewhere that's copied and all its content never used, i.e "dangling".

NilsLankila commented:
try to add 

```d
    override Dsymbol syntaxCopy(Dsymbol s)
    {
        return this;
    }
```

in `class` Package

andralex commented:
@NilsLankila I experimented a bit with that override. Didn't have time to collect the detailed logs, but based on the memory usage alone there seems to be no measurable difference.

NilsLankila commented:
Then certainly my guess about syntax copy is wrong.

andralex committed:
Cosmetic

andralex commented:
I'm unclear how to go about debug definition. So there are three possibilities:

- no debugging (most often)
- debug with summaries, prints one line per call site at program exit (second most often)
- debug verbose, prints one line in real time per call to Pool functions (seldom used, mostly for debugging crashes)

I'm not sure what's the nicest way to go about that. Currently there are two debug symbols, `Pool` and `PoolVerbose`. But that's a bit awkward. Thoughts?

UplinkCoder commented:
This introduced a regression in the compilation of one of our (symmetry's) projects.

UplinkCoder commented:
there is a linked list inside the template instances which needs to be unlinked once you dispose an object.

andralex commented:
>This introduced a regression in the compilation of one of our (symmetry's) projects.

To debug a use-after-dispose problem, define `debug = Pool` and `debug = PoolVerbose`. Usually the last lines output just before the crash are the calls to `dispose()` that need to be eliminated. A better method would be to have a `bool disposed;` field in the object and `assert` on it, but that would be justifiable only if the pool gets some real usage.

>there is a linked list inside the template instances which needs to be unlinked once you dispose an object.

If (after fixing the culprit) you could do that and compact memory, that would be awesome.

UplinkCoder commented:
> > This introduced a regression in the compilation of one of our (symmetry's) projects.
> 
> To debug a use-after-dispose problem, define `debug = Pool` and `debug = PoolVerbose`. Usually the last lines output just before the crash are the calls to `dispose()` that need to be eliminated. A better method would be to have a `bool disposed;` field in the object and `assert` on it, but that would be justifiable only if the pool gets some real usage.
> 
> > there is a linked list inside the template instances which needs to be unlinked once you dispose an object.
> 
> If (after fixing the culprit) you could do that and compact memory, that would be awesome.

First of all this is your code, and therefore it is your responsibility to address issues caused by it.

The reason this memory corruption happens is because
  - dmd assumes infinite object lifetime (at least in AST context)
     - A pointer (and class refs are pointers) is either null, or does point to valid memory.
       Anything which violates that assumption is going to cause memory corruption bugs.
       Those are by nature hard to find, and it is almost impossible to assert that there are none.
    Therefore:
       the only proper solution, as far as I can see, is to give on the idea of reusing memory, in circumstances which concern the AST.
      Another solution would be to re-architect how dmd manages memory completely but I don't see how that would not break even more code.

If my explanation of the problem was unclear or lacking, I'll be happy to clarify upon request.

UplinkCoder commented:
@andralex I will have to amend my testcase, as the particular reduction linked in the issue does not succeed even if this change is reverted.

The testcase will be amdended.

Also your suggested debug method generates this file: https://gist.github.com/UplinkCoder/06ff78238b74a491d097a7308f3529ad
which I am not quite sure what to do with.

ibuclaw commented:
> @andralex I will have to amend my testcase, as the particular reduction linked in the issue does not succeed even if this change is reverted.
> 
> The testcase will be amdended.
> 
> Also your suggested debug method generates this file: https://gist.github.com/UplinkCoder/06ff78238b74a491d097a7308f3529ad
> which I am not quite sure what to do with.

It probably means that in this snippet.

https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L2957-L2969

The call to `deduceFunctionTemplateMatch` is not without side effects and it stows away the template instance somewhere only for it to be referenced after it has been disposed.

andralex commented:
I'm okay with reverting this, right now we experience the disadvantages and no advantage of object pooling. Also, problem is future maintenance may start reusing objects that were `dispose()`d of so keeping that in mind is burdensome.

Sadly I could not revert automatically, not sure what happens.
